### DISCUSSION TITLE
Add a warning if a student grades a gateway quiz without answering all questions

### DISCUSSION THREAD

**Discussion by dlglin** [openwebwork/webwork2#1612] [Ideas]
I'd like to have an alert pop up if a student clicks "Grade test" on a gateway quiz before they have answered all of the questions reminding them to answer the rest of the questions.  I'm trying to think through all of the scenarios for this.

Easier implementation: any time a student clicks "Grade test" they get an alert saying something like "This will use one of your submissions.  Please make sure that you have answered all questions before continuing".  The problem is that this may not be relevant in many scenarios (e.g. unlimited attempts allowed).

More involved implementation: write some logic to determine whether to display an alert.  Things to possibly check:

- Are there any empty answer boxes
- If the quiz allows unlimited attempts, then suppress the alerts.
- If it is near the deadline, maybe don't warn about empty answer boxes (since students will want to submit an incomplete exam at that point)?
- If the quiz is split into multiple pages, can js determine if questions on other pages remain unanswered?

There are probably additional scenarios that I'm not thinking of, which is why the first option is presented.  The question is, could this be done with a reasonable amount of effort in a way that is beneficial in most cases?

**Comment by drgrice1** [comment]
I am planning on implementing something like this. Should be easy.

**Comment by drgrice1** [comment]
One thing that I disagree with you slightly on is the matter of not showing the dialog if the student has answered all questions.  I am thinking of the situation where a student has answered all questions, and thinks okay now I want to save my answers and then look over all of the answers one more time before I am done.  Of course if that was the last submission the student will not be able to change answers subsequently in the final look through.

In general  am wondering what does it hurt to just show an alert in any case except that of the infinite number of submissions and perhaps not in the last 90 seconds (when there are other alerts already telling the student to click "Grade Test".  The alert should state the number of submissions that remain and that once all submissions are used no changes to answers will be allowed.  If this is the last submission it should state that.

Javascript can determine if answers have been entered for all questions on the page just as well as the GatewayQuiz.pm module itself does, and using the same method.

**Comment by dlglin** [comment]
I'm fine with the alert always being there.  I was just trying to brainstorm scenarios where it may be more of a hindrance than a benefit, but I'm having a hard time convincing myself that showing it more often than necessary would be a problem.

**Comment by drgrice1** [comment]
I thought that it would be easy to detect missing answers in javascript, but after looking at it closer I see that it is not.  The gateway quiz module adds hidden fields for answers that are not on the page, but it only does so if the field has an answer in it.  So any unanswered question or part of a question would not be present.  Furthermore, even the gateway quiz module does not not what answers the questions that are not on the page have.  So the conclusion is that it actually isn't possible to reliably determine that there are unanswered questions in javascript.

I put in a pull request that implements this (without answer checks).  See openwebwork/webwork2#1613.
